include/xtensor/xcontainer.hpp

```diff
@@ -260,13 +266,11 @@ namespace xt
     }
 
     template <class D>
-    inline auto xcontainer<D>::data_offset(const xindex& index) const noexcept -> size_type
+    template <class It>
+    inline auto xcontainer<D>::offset(It first, It last) const noexcept -> size_type
     {
-        // VS2015 workaround : index.begin() + index.size() - m_strides.size()
-        // doesn't compile
-        auto iter = index.begin();
-        iter += index.size() - m_strides.size();
-        return std::inner_product(m_strides.begin(), m_strides.end(), iter, size_type(0));
+        first += std::distance(first, last) - m_strides.size();
+        return std::inner_product(m_strides.begin(), m_strides.end(), first, size_type(0));
```

Comment:
I think we should be consistent here: either we support random access iterators only and we can replace `std::distance` with `operator-` or we support bidirectional iterators and we must use `std::advance` instead of `operator+`